src/gui/widgets/subtitlelistwidget.cpp

```diff
@@ -470,15 +483,41 @@ void SubtitleListWidget::handleSecondaryTrackChange(int64_t sid)
     m_secondaryLock.unlock();
 }
 
-void SubtitleListWidget::addPrimarySubtitle(const QString &subtitle,
-                                            const double   start,
-                                            const double   end,
-                                            const double   delay)
+void SubtitleListWidget::updatePrimarySubtitle(const QString &subtitle,
+                                               const double   start,
+                                               const double   end,
+                                               const double   delay)
 {
     if (!m_primaryLock.try_lock())
     {
         return;
     }
+    // if parsing has been completed
+    // assume entry is in list and only look up
+    if (m_didParsePrimarySubs)
+    {
+        double timePos = GlobalMediator::getGlobalMediator()
+                         ->getPlayerAdapter()->getTimePos();
+        auto it = --m_seenPrimarySubs
+                  .upperBound(timePos + TIME_DELTA - delay);
+        QStack<QTableWidgetItem *> visibleSubs;
+        visibleSubs.push(*it);
+        while (it-- != m_seenPrimarySubs.begin() &&
+               m_timesPrimarySubs[it.value()]->end >
+               timePos + TIME_DELTA - delay)
```

Comment:
So this implementation stops iterating backwards by start time when there is a subtitle whose end is earlier than the current playback time. There is a case where it breaks:

```
----------------- Sub1---------------------------
                 --------Sub2---------
                                            ----------Sub3-----------
                                               ^timePos
```
In this case, Sub1 is still visible on screen, but because Sub2 both started later and ended earlier than it, we have stopped searching. I have considered allowing `N_MAX_CONTINUES` searches that resume after skipping lines like Sub2, but feel that it is a rare case that does not justify overhead for each call.

Reply:
I can't think of any good solutions to this problem that aren't O(n) each time the method is called, but there is a clever solution.

That way mpv flattens subtitles, each different subtitle will be put on a new line. Using that, QMultiHash can be used that maps lines of subtitles to `SubtitleInfo *`. From there we can just check if that `SubtitleInfo`'s start and end times are between the current time and then select the corresponding item if it is. This algorithm should be average O(number of newlines in an mpv subtitle).